async_cache.go

```diff
@@ -0,0 +1,102 @@
+package cache
+
+import (
+	"fmt"
+	"sync"
+)
+
+//Error Codes for keystates
+var (
+	ErrInvalidKey         = fmt.Errorf("found invalid key")
+	ErrItemNotPresent     = fmt.Errorf("item not present")
+	ErrKeyStateNotPresent = fmt.Errorf("keystate datastructure not initialized")
+)
+
+//keyState structure
+type keyStates struct {
+	keyState map[string]state //states{INPROCESS,DONE,INVALID,NOPRESENT}
+	mu       *sync.RWMutex    //RWmutex
+}
+
+type state int
+
+//states declared in constants
+const (
+	INPROCESS state = iota
+	DONE
+	INVALID
+	NOPRESENT
+)
+
+var ks keyStates
+
+//initializing ks keyStates
+func init() {
+	ks.keyState = make(map[string]state)
+	ks.mu = &sync.RWMutex{}
+}
+
+//Set status/state with respective key in keyState
+func (ks *keyStates) Set(key string, status state) (bool, error) {
```

Comment:
as status is enough to understand the calling function if there is any issue. No need to return the error, please remove it.

Reply:
done